### PR TITLE
sstable: reorganize sstable buffers and structs

### PR DESCRIPTION
**Pull request by bananabrick**
This pr aims to serve as a first step towards unifying the parallel compression vs
on parallel compression code paths, as suggested in this pr:
#1383 (review).

This pr organizes the buffers in the Writer in a way which makes it easier to reason
about the ownership of buffers/slices.

Future prs will add, a writeQueue, a compressionQueue, a mechanism to schedule parallel
compression based on cpu utilization.

### CODE REVIEW

**Review by bananabrick — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1442)* status: 0 of 1 files reviewed, 1 unresolved discussion (waiting on @itsbilal, @jbowens, and @petermattis)

___
*[sstable/writer.go, line 609 at r1](https://reviewable.io/reviews/cockroachdb/pebble/1442#-MsmAWqJ8crOt7FNHsAt:-MsmAWqJ8crOt7FNHsAu:byev94r) ([raw file](https://github.com/cockroachdb/pebble/blob/6a59c79295cc844679eba90896c8f78d5ef9b042/sstable/writer.go#L609)):*
> ```Go
> 	}
> 	w.writeQueue.Add(writeTask)
> 	// Try and acquire a new passableWriterState for future use. Note that since we've only initialized
> ```

Unfortunately, this part of the code where we add a `writeQueueTask` to the queue and block causes a `~10%` perf regression in ycsb/F.

The options are to:
1. Write a more efficient queue which would work for both no parallel compression vs parallel compression. (not sure if this is possible).
2. Have a more efficient queue just for no parallel compression.

If we use 2, the codepaths will be less unified, but it'll probably be easier than 1.

**Review by bananabrick — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1442)* status: 0 of 1 files reviewed, 1 unresolved discussion (waiting on @itsbilal, @jbowens, and @petermattis)

___
*[sstable/writer.go, line 609 at r1](https://reviewable.io/reviews/cockroachdb/pebble/1442#-MsmAWqJ8crOt7FNHsAt:-MsmE2Sv9b7-fSNwOjro:b-gvj9c) ([raw file](https://github.com/cockroachdb/pebble/blob/6a59c79295cc844679eba90896c8f78d5ef9b042/sstable/writer.go#L609)):*
<details><summary><i>Previously, bananabrick (Arjun Nair) wrote…</i></summary><blockquote>

Unfortunately, this part of the code where we add a `writeQueueTask` to the queue and block causes a `~10%` perf regression in ycsb/F.

The options are to:
1. Write a more efficient queue which would work for both no parallel compression vs parallel compression. (not sure if this is possible).
2. Have a more efficient queue just for no parallel compression.

If we use 2, the codepaths will be less unified, but it'll probably be easier than 1.
</blockquote></details>

For example, we could have a `writeQueue` interface, and use different implementations for parallel compression vs no parallel compression. I'm confident that I can write a `writeQueue` implementation which will require no synchronization for the no parallel compression case.

**Review by bananabrick — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1442)* status: 0 of 1 files reviewed, 1 unresolved discussion (waiting on @itsbilal, @jbowens, and @petermattis)

___
*[sstable/writer.go, line 609 at r1](https://reviewable.io/reviews/cockroachdb/pebble/1442#-MsmAWqJ8crOt7FNHsAt:-Mt5Hp6j-jNU5858jp1_:bhtxgxk) ([raw file](https://github.com/cockroachdb/pebble/blob/6a59c79295cc844679eba90896c8f78d5ef9b042/sstable/writer.go#L609)):*
<details><summary><i>Previously, bananabrick (Arjun Nair) wrote…</i></summary><blockquote>

For example, we could have a `writeQueue` interface, and use different implementations for parallel compression vs no parallel compression. I'm confident that I can write a `writeQueue` implementation which will require no synchronization for the no parallel compression case.
</blockquote></details>

Option 2 ended up working. Pushed a 2nd commit with the change.

**Review by jbowens — COMMENTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1442)* status: 0 of 1 files reviewed, 8 unresolved discussions (waiting on @bananabrick, @itsbilal, and @petermattis)

___
*[sstable/writer.go, line 775 at r1](https://reviewable.io/reviews/cockroachdb/pebble/1442#-Mt57H9IAG8xgs-E3joQ:-Mt57H9IAG8xgs-E3joR:b81advv) ([raw file](https://github.com/cockroachdb/pebble/blob/6a59c79295cc844679eba90896c8f78d5ef9b042/sstable/writer.go#L775)):*
> ```Go
> 
> func compressAndChecksum(
> 	b []byte, compression Compression, compressedBuf *[]byte, trailerBuf *[blockHandleLikelyMaxLen]byte,
> ```

does `compressedBuf` need to be a pointer? If compression is worthwhile, we return out the compressed byte slice through the return value. We can mutate the slice's underlying array without the extra pointer indirection.
___
*[sstable/writer.go, line 814 at r1](https://reviewable.io/reviews/cockroachdb/pebble/1442#-Mt57pq47EjuqkehlDA-:-Mt57pq47EjuqkehlDA0:b498md9) ([raw file](https://github.com/cockroachdb/pebble/blob/6a59c79295cc844679eba90896c8f78d5ef9b042/sstable/writer.go#L814)):*
> ```Go
> 
> func (w *Writer) writeBlockPostCompression(
> 	toWrite []byte, blockTrailerBuf [blockHandleLikelyMaxLen]byte,
> ```

Taking the `blockTrailerBuf` parameter as an array will cause Go to allocate (probably on the stack) `blockHandleLikelyMaxLen` bytes and copy the argument's contents into it. We could take a slice of the array though, passing in `w.curWriterState.tmp[:]`.

nit: s/`toWrite`/`block`/?
___
*[sstable/writer.go, line 148 at r2](https://reviewable.io/reviews/cockroachdb/pebble/1442#-Mt9-aFA7qVjSNXEJKti:-Mt9-aFA7qVjSNXEJKtj:b-qf2571) ([raw file](https://github.com/cockroachdb/pebble/blob/5155d6cde2fdcb0dc26b56795c325a3c9669f291/sstable/writer.go#L148)):*
> ```Go
> 	filter filterWriter
> 
> 	topLevelIndexBlock blockWriter
> ```

nit: should this be alongside the other `blockWriter`s?
___
*[sstable/writer.go, line 152 at r2](https://reviewable.io/reviews/cockroachdb/pebble/1442#-Mt9-lxvAS53KUb3ap2O:-Mt9-lxvAS53KUb3ap2P:b-v3qsik) ([raw file](https://github.com/cockroachdb/pebble/blob/5155d6cde2fdcb0dc26b56795c325a3c9669f291/sstable/writer.go#L152)):*
> the compaction/Writer goroutine

nit: This phrasing is a little ambiguous, although I'm not sure what to call it. The "sstable-constructing goroutine"? The "Writer client goroutine"?
___
*[sstable/writer.go, line 185 at r2](https://reviewable.io/reviews/cockroachdb/pebble/1442#-Mt8zp_q4SCxPZMXE61Q:-Mt8zp_q4SCxPZMXE61R:bw0gmj6) ([raw file](https://github.com/cockroachdb/pebble/blob/5155d6cde2fdcb0dc26b56795c325a3c9669f291/sstable/writer.go#L185)):*
> ```Go
> 	compressedBuf []byte
> 
> 	dataBlockWriter *blockWriter
> ```

nit: how about `dataBlock` for parity with `Writer.{indexBlock,rangeDelBlock,rangeKeyBlock}`
___
*[sstable/writer.go, line 230 at r2](https://reviewable.io/reviews/cockroachdb/pebble/1442#-Mt92SWg6KQFkcdcplEa:-Mt92SWg6KQFkcdcplEb:b-jefgxk) ([raw file](https://github.com/cockroachdb/pebble/blob/5155d6cde2fdcb0dc26b56795c325a3c9669f291/sstable/writer.go#L230)):*
> ```Go
> 	// passableWriterState is temporary.
> 	task.writerState.err = err
> 	task.writerState.blockHandle = bh
> ```

Passing information back through the `writerStateBuffer.put` is unexpected and hard to follow the control. I think we should address this and remove `err` and `blockHandle` in this PR.
___
*[sstable/writer.go, line 238 at r2](https://reviewable.io/reviews/cockroachdb/pebble/1442#-Mt91UVD0a51CjyQq49N:-Mt91UVD0a51CjyQq49O:bdee95i) ([raw file](https://github.com/cockroachdb/pebble/blob/5155d6cde2fdcb0dc26b56795c325a3c9669f291/sstable/writer.go#L238)):*
> ```Go
> 
> // passableStateBuffer initially owns every initialized passableWriterState.
> type passableStateBuffer interface {
> ```

I think we should replace this with a `sync.Pool`. It's easier to read, since it's idiomatic, and it's likely to be more performant once we add concurrency than a hand-rolled pool.

**Review by bananabrick — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1442)* status: 0 of 1 files reviewed, 5 unresolved discussions (waiting on @itsbilal, @jbowens, and @petermattis)

___
*[sstable/writer.go, line 775 at r1](https://reviewable.io/reviews/cockroachdb/pebble/1442#-Mt57H9IAG8xgs-E3joQ:-Mt9JYmJAX_FvtwGKEQ8:btylami) ([raw file](https://github.com/cockroachdb/pebble/blob/6a59c79295cc844679eba90896c8f78d5ef9b042/sstable/writer.go#L775)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

does `compressedBuf` need to be a pointer? If compression is worthwhile, we return out the compressed byte slice through the return value. We can mutate the slice's underlying array without the extra pointer indirection.
</blockquote></details>

I was making use of the fact that `compressedBuf` is a pointer in this code:
```
	if blockType != noCompressionBlockType && cap(compressed) > cap(*compressedBuf) {
		*compressedBuf = compressed[:cap(compressed)]
	}
```
In this case, the `compressed` slice was newly allocated, because `compressedBuf`, wasn't large enough. Are you suggesting moving this `if` block outside the `compressAndChecksum` function?
___
*[sstable/writer.go, line 814 at r1](https://reviewable.io/reviews/cockroachdb/pebble/1442#-Mt57pq47EjuqkehlDA-:-Mt9J8X3BmEdGVOtSV90:b-896fix) ([raw file](https://github.com/cockroachdb/pebble/blob/6a59c79295cc844679eba90896c8f78d5ef9b042/sstable/writer.go#L814)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

Taking the `blockTrailerBuf` parameter as an array will cause Go to allocate (probably on the stack) `blockHandleLikelyMaxLen` bytes and copy the argument's contents into it. We could take a slice of the array though, passing in `w.curWriterState.tmp[:]`.

nit: s/`toWrite`/`block`/?
</blockquote></details>

Done.
___
*[sstable/writer.go, line 152 at r2](https://reviewable.io/reviews/cockroachdb/pebble/1442#-Mt9-lxvAS53KUb3ap2O:-Mt9HFip0SdkOlp1Fq8V:b-lh5jyk) ([raw file](https://github.com/cockroachdb/pebble/blob/5155d6cde2fdcb0dc26b56795c325a3c9669f291/sstable/writer.go#L152)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

> the compaction/Writer goroutine

nit: This phrasing is a little ambiguous, although I'm not sure what to call it. The "sstable-constructing goroutine"? The "Writer client goroutine"?
</blockquote></details>

Writer client sgtm.
___
*[sstable/writer.go, line 230 at r2](https://reviewable.io/reviews/cockroachdb/pebble/1442#-Mt92SWg6KQFkcdcplEa:-Mt9aYcCAU57FIx9OPLd:b-896fix) ([raw file](https://github.com/cockroachdb/pebble/blob/5155d6cde2fdcb0dc26b56795c325a3c9669f291/sstable/writer.go#L230)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

Passing information back through the `writerStateBuffer.put` is unexpected and hard to follow the control. I think we should address this and remove `err` and `blockHandle` in this PR.
</blockquote></details>

Done.
___
*[sstable/writer.go, line 238 at r2](https://reviewable.io/reviews/cockroachdb/pebble/1442#-Mt91UVD0a51CjyQq49N:-Mt9HWzE5W70ZVcHqsOw:bl9czx8) ([raw file](https://github.com/cockroachdb/pebble/blob/5155d6cde2fdcb0dc26b56795c325a3c9669f291/sstable/writer.go#L238)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

I think we should replace this with a `sync.Pool`. It's easier to read, since it's idiomatic, and it's likely to be more performant once we add concurrency than a hand-rolled pool.
</blockquote></details>

I made it an interface, so that the no parallel compression case will have a faster buffer. I could use a sync.Pool for both parallel vs no parallel compression, but I fear it will cause a mild perf hit for the no parallel compression case.

For the parallel compression case, I plan on using a fixed size buffer of length `CompressionConcurrency`. We shouldn't need any more than that, since this buffer exists per Writer. But I can also use `sync.Pool`.

**Review by jbowens — COMMENTED** [review]
I'm still having a little trouble piecing this together, and I'm not sure as written the write queue unifies quite far enough.

I think it'd be nice to even further restrict this first PR to just reorganizing the buffers without introducing any queues. I think one goal of the code structure is that he Writer client goroutine should be able to compress and enqueue its own blocks if the queue is full.

I'm thinking something like:

```
type blockBuf struct {
	// buffers, reused between blocks.
	*blockWriter
	compressedBuf []byte
	trailer       [blockTrailerLen]byte
	xxHasher      *xxhash.Digest

	// block request state
	uncompressed  []byte
	compressed    []byte
	compression   Compression
	checksumType  ChecksumType
}

func (bb *blockBuf) compress() error {}

func (w *Writer) writeCompressedBlock(bb *blockBuf) (BlockHandle, error) {}
```

One `blockBuf` would handle the entire lifecycle of a block.

The `Writer` type would have one `blockBuf` field embedded for writing non-data blocks (which I think will only ever be constructed sequentially?). The `Writer.block` field would become a  separate`*blockBuf`. In the parallel world these `*blockBuf`s may be pooled. Does this make sense?

----

Sorry this has been such a slog. SStable construction doesn't parallelize easily and the sstable writer is so foundational that it's very important we get this right.

> Reviewed 1 of 2 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1442#-:-MtFOkAUEhcB04WOjACT:b-ggb02a)* status: 0 of 3 files reviewed, 5 unresolved discussions (waiting on @bananabrick, @itsbilal, @jbowens, and @petermattis)

___
*[sstable/writer.go, line 775 at r1](https://reviewable.io/reviews/cockroachdb/pebble/1442#-Mt57H9IAG8xgs-E3joQ:-MtEYp371sZsfrIwEgsm:b7rlle6) ([raw file](https://github.com/cockroachdb/pebble/blob/6a59c79295cc844679eba90896c8f78d5ef9b042/sstable/writer.go#L775)):*
<details><summary><i>Previously, bananabrick (Arjun Nair) wrote…</i></summary><blockquote>

I was making use of the fact that `compressedBuf` is a pointer in this code:
```
	if blockType != noCompressionBlockType && cap(compressed) > cap(*compressedBuf) {
		*compressedBuf = compressed[:cap(compressed)]
	}
```
In this case, the `compressed` slice was newly allocated, because `compressedBuf`, wasn't large enough. Are you suggesting moving this `if` block outside the `compressAndChecksum` function?
</blockquote></details>

Yeah, lifting that into the caller(s) is preferable, imo, because it's a lot clearer whether appropriate synchronization is in place.
___
*[sstable/writer.go, line 293 at r5](https://reviewable.io/reviews/cockroachdb/pebble/1442#-MtEQO6VEhVClHQ2OFsT:-MtEQO6VEhVClHQ2OFsU:b-qse7xa) ([raw file](https://github.com/cockroachdb/pebble/blob/f42ecb9e95ba9d8308d20e89a8b24af7e5e9dc93/sstable/writer.go#L293)):*
> ```Go
> 	if len(w.errCh) > 0 {
> 		w.err = <-w.errCh
> 	}
> ```

nit: this might be a micro optimization (although this _is_ the sstable-writing hot path) but I expect

```
select {
case w.err <- w.errCh:
default:
}
```
to be slightly faster, because it avoids reading (and synchronizing with)  the channel internal state twice.
___
*[sstable/writer.go, line 163 at r9](https://reviewable.io/reviews/cockroachdb/pebble/1442#-MtEeELmC4UuHDIjgmzF:-MtEeELmC4UuHDIjgmzG:b-utbqwz) ([raw file](https://github.com/cockroachdb/pebble/blob/c166b2294c5f3543dd9addb73e46536512d08cb3/sstable/writer.go#L163)):*
> ```Go
> 	writeQueue writeQueue
> 
> 	checksummer checksummer
> ```

`checksummer.xxHasher` isn't safe for concurrent use — should there be a `checksummer` field in `passableWriterState`?

**Review by bananabrick — COMMENTED** [review]
I can work on restricting this pr even further.

However, if we want to unify parallel compression vs no parallel compression, even further, we would have to use a single queue implementation for both. Parallel compression will require a queue with some synchronization, which slows down the perf of the no parallel compression case.

I'll think about simpler write/compression queue ideas which can be used for both.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1442#-:-MtFgWov8AnH0ok2X5im:b-850634)* status: 0 of 3 files reviewed, 5 unresolved discussions (waiting on @bananabrick, @itsbilal, @jbowens, and @petermattis)

**Review by bananabrick — COMMENTED** [review]
Although, maybe we can have the code unified, without using a queue for the no parallel compression case.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1442#-:-MtFjTa21H5w01RqWk5u:b-rx3god)* status: 0 of 3 files reviewed, 5 unresolved discussions (waiting on @bananabrick, @itsbilal, @jbowens, and @petermattis)

**Review by itsbilal — COMMENTED** [review]
I think what Jackson is hinting at (and what I'd encourage too) is to try to keep all the logic that's currently in `compressAndChecksum` confined to a `blockBuf.Compress`, that could optionally be called by either the writer goroutine (and always by the writer goroutine if there's no parallel compression), or a separate goroutine. The synchronization bits that you're currently putting into a `writeQueue` implementation could probably follow that design no problem, except that you may no longer need a `defaultWriteQueue` for the non-parallel case; that logic could just be in `maybeFlush`.

> Reviewed 1 of 2 files at r7, 1 of 2 files at r8, 1 of 1 files at r9.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1442#-:-MtJzULz2bdrw6ut3lin:bg9xq6o)* status: all files reviewed (commit messages unreviewed), 9 unresolved discussions (waiting on @bananabrick, @jbowens, and @petermattis)

___
*[sstable/writer.go, line 775 at r1](https://reviewable.io/reviews/cockroachdb/pebble/1442#-Mt57H9IAG8xgs-E3joQ:-MtJvy_GD0IK6NMcOBnB:b-v9dhle) ([raw file](https://github.com/cockroachdb/pebble/blob/6a59c79295cc844679eba90896c8f78d5ef9b042/sstable/writer.go#L775)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

Yeah, lifting that into the caller(s) is preferable, imo, because it's a lot clearer whether appropriate synchronization is in place.
</blockquote></details>

I agree with Jackson, plus that gets us to avoid the use of a slice pointer here.
___
*[sstable/writer.go, line 163 at r6](https://reviewable.io/reviews/cockroachdb/pebble/1442#-MtF4xu1C5Eycwnn4ueW:-MtF4xu20B89mRkPT6BI:bi7z002) ([raw file](https://github.com/cockroachdb/pebble/blob/b8d3bab0cabc9e45a3937bdb8fcc4ec1b4991e64/sstable/writer.go#L163)):*
> ```Go
> 	writeQueue writeQueue
> 
> 	checksummer checksummer
> ```

Nit: no need for the empty lines in between.
___
*[sstable/writer.go, line 163 at r9](https://reviewable.io/reviews/cockroachdb/pebble/1442#-MtEeELmC4UuHDIjgmzF:-MtJxSxnC61YxEki1eJu:b-9wytoc) ([raw file](https://github.com/cockroachdb/pebble/blob/c166b2294c5f3543dd9addb73e46536512d08cb3/sstable/writer.go#L163)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

`checksummer.xxHasher` isn't safe for concurrent use — should there be a `checksummer` field in `passableWriterState`?
</blockquote></details>

Either that, or the `writeQueue` could have a slice of checksummers - one for each worker (and just one for the defaultWriteQueue).
___
*[sstable/writer.go, line 294 at r9](https://reviewable.io/reviews/cockroachdb/pebble/1442#-MtJQ1HuDu8qC4p4VHT6:-MtJQ1Hv2YnfPqZ284CV:b-9itr09) ([raw file](https://github.com/cockroachdb/pebble/blob/c166b2294c5f3543dd9addb73e46536512d08cb3/sstable/writer.go#L294)):*
> ```Go
> }
> 
> func (w *Writer) fetchEncounteredError() error {
> ```

nit: this could just be `Error()`.
___
*[sstable/writer.go, line 808 at r9](https://reviewable.io/reviews/cockroachdb/pebble/1442#-MtJv1Fa452NDR7pMutD:-MtJv1Fb3A5wGcPDeS8w:bu4gmk4) ([raw file](https://github.com/cockroachdb/pebble/blob/c166b2294c5f3543dd9addb73e46536512d08cb3/sstable/writer.go#L808)):*
> ```Go
> 
> 	// We're passing ownership of prevWriterState to the writeQueue.
> 	writeTask := &writeQueueTask{
> ```

This would cause a heap alloc right? We should just get this off of a pool instead of doing the alloc.
___
*[sstable/writer.go, line 966 at r9](https://reviewable.io/reviews/cockroachdb/pebble/1442#-MtJwI5U-KjkrJicRakf:-MtJwI5U-KjkrJicRakg:b4vi6ol) ([raw file](https://github.com/cockroachdb/pebble/blob/c166b2294c5f3543dd9addb73e46536512d08cb3/sstable/writer.go#L966)):*
> ```Go
> 
> func compressAndChecksum(
> 	b []byte, compression Compression, compressedBuf *[]byte, trailerBuf *[blockHandleLikelyMaxLen]byte,
> ```

I'm not seeing `trailerBuf` get resliced or something, so why not just change the type to `[]byte` instead and have the caller pass in `trailerBuf[:]`? Makes the code cleaner, plus any of the out-of-range indexing calls would still panic.

**Review by bananabrick — COMMENTED** [review]
I restricted this pr to only include the buffers. I didn't make the `blockBuf` a receiver argument for compression, and block writes. This is because a `blockBuf`, and a `dataBlockBuf`, both need to use those functions. The reason a `blockBuf`, and a `dataBlockBuf`, are separate types is because the `blockBuf` is used to write non data blocks, and there are many types of non data blocks.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1442#-:-MtdD7cX7UxuO0L7PglG:b-f6o9ci)* status: 1 of 3 files reviewed, 6 unresolved discussions (waiting on @itsbilal, @jbowens, and @petermattis)

___
*[sstable/writer.go, line 775 at r1](https://reviewable.io/reviews/cockroachdb/pebble/1442#-Mt57H9IAG8xgs-E3joQ:-MtiH20y225qL94yZvZx:b-f1x5gn) ([raw file](https://github.com/cockroachdb/pebble/blob/6a59c79295cc844679eba90896c8f78d5ef9b042/sstable/writer.go#L775)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

I agree with Jackson, plus that gets us to avoid the use of a slice pointer here.
</blockquote></details>

I changed this function a bit, so `compressedBuf` is no longer being updated through a pointer. Lifting the `compressedBuf` resizing out of this function requires us to know whether an allocation happened, and it didn't seem worthwhile to return that information.
___
*[sstable/writer.go, line 163 at r9](https://reviewable.io/reviews/cockroachdb/pebble/1442#-MtEeELmC4UuHDIjgmzF:-Mtd-nmv6zfHzRnXrBje:b-896fix) ([raw file](https://github.com/cockroachdb/pebble/blob/c166b2294c5f3543dd9addb73e46536512d08cb3/sstable/writer.go#L163)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

Either that, or the `writeQueue` could have a slice of checksummers - one for each worker (and just one for the defaultWriteQueue).
</blockquote></details>

Done.
___
*[sstable/writer.go, line 808 at r9](https://reviewable.io/reviews/cockroachdb/pebble/1442#-MtJv1Fa452NDR7pMutD:-Mtd8EMP4e3-5nmMOvul:b-896fix) ([raw file](https://github.com/cockroachdb/pebble/blob/c166b2294c5f3543dd9addb73e46536512d08cb3/sstable/writer.go#L808)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

This would cause a heap alloc right? We should just get this off of a pool instead of doing the alloc.
</blockquote></details>

Done.
___
*[sstable/writer.go, line 966 at r9](https://reviewable.io/reviews/cockroachdb/pebble/1442#-MtJwI5U-KjkrJicRakf:-MtdAO-N2au8FTF0XJ75:b-896fix) ([raw file](https://github.com/cockroachdb/pebble/blob/c166b2294c5f3543dd9addb73e46536512d08cb3/sstable/writer.go#L966)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

I'm not seeing `trailerBuf` get resliced or something, so why not just change the type to `[]byte` instead and have the caller pass in `trailerBuf[:]`? Makes the code cleaner, plus any of the out-of-range indexing calls would still panic.
</blockquote></details>

Done.

**Review by jbowens — COMMENTED** [review]
Looking great!

The separate `dataBlockBuf` and `blockBuf` types make sense :+1: 

> I didn't make the `blockBuf` a receiver argument for compression.

It might (eventually) be useful to add a `compressed` field to the `dataBlockBuf` type, since the `dataBlockBuf`'s lifecycle is already tied to the output of `compressAndChecksum` (because the returned `[]byte` may point directly into `dataBlockBuf.blockWriter`'s buffer).

> Reviewed 1 of 1 files at r9, 1 of 2 files at r10, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1442#-:-MtieQ2W88KDwFeMcHpl:b-rf6c30)* status: 2 of 3 files reviewed, 7 unresolved discussions (waiting on @bananabrick, @itsbilal, @jbowens, and @petermattis)

___
*[sstable/writer.go, line 161 at r10](https://reviewable.io/reviews/cockroachdb/pebble/1442#-MtidRyA1741Wk30jbNZ:-MtidRyB3mRSwaruqSOy:barmo8r) ([raw file](https://github.com/cockroachdb/pebble/blob/0b84911cabe1b57ece33aae3032be47e5f0719ef/sstable/writer.go#L161)):*
> ```Go
> 	// blockBuf consists of the state which is owned by and used by the Writer client
> 	// goroutine.
> 	blockBuf *blockBuf
> ```

looks like `Writer.blockBuf` doesn't need to be a pointer.
___
*[sstable/writer.go, line 182 at r10](https://reviewable.io/reviews/cockroachdb/pebble/1442#-MtifjsfCNW9tG1hIQyC:-MtifjsfCNW9tG1hIQyD:b-ylse53) ([raw file](https://github.com/cockroachdb/pebble/blob/0b84911cabe1b57ece33aae3032be47e5f0719ef/sstable/writer.go#L182)):*
> ```Go
> // A dataBlockBuf is owned by the Writer client goroutine. Once {compression, write}Queues
> // are added, blockBuf could also be owned by the {compression, write}Queue goroutines.
> // A dataBlockBuf has all the state required to compress and write a data block to disk.
> ```

I found the discussion of ownership a little muddled here. How about:

```
// A dataBlockBuf holds all the state required to compress and write a data block to disk.
// A dataBlockBuf begins its lifecycle owned by the Writer client goroutine. The Writer
// client goroutine adds keys to the sstable, writing directly into a dataBlockBuf's blockWriter
// until the block is full. Once a dataBlockBuf's block is full, the dataBlockBuf may be passed
// to other goroutines for compression and file I/O.
```
___
*[sstable/writer.go, line 185 at r10](https://reviewable.io/reviews/cockroachdb/pebble/1442#-Mtif2yR9bR68fitgjDs:-Mtif2yR9bR68fitgjDt:b15dklh) ([raw file](https://github.com/cockroachdb/pebble/blob/0b84911cabe1b57ece33aae3032be47e5f0719ef/sstable/writer.go#L185)):*
> ```Go
> type dataBlockBuf struct {
> 	blockBuf
> 	dataBlock *blockWriter
> ```

It looks like we can embed the `blockWriter` and avoid the pointer indirection.

**Review by bananabrick — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1442)* status: 2 of 3 files reviewed, 6 unresolved discussions (waiting on @itsbilal, @jbowens, and @petermattis)

___
*[sstable/writer.go, line 161 at r10](https://reviewable.io/reviews/cockroachdb/pebble/1442#-MtidRyA1741Wk30jbNZ:-MtjhMeE2EdkWUSXvNNb:b-896fix) ([raw file](https://github.com/cockroachdb/pebble/blob/0b84911cabe1b57ece33aae3032be47e5f0719ef/sstable/writer.go#L161)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

looks like `Writer.blockBuf` doesn't need to be a pointer.
</blockquote></details>

Done.
___
*[sstable/writer.go, line 182 at r10](https://reviewable.io/reviews/cockroachdb/pebble/1442#-MtifjsfCNW9tG1hIQyC:-Mtjgsqr2km8-74WRMYb:b604hgk) ([raw file](https://github.com/cockroachdb/pebble/blob/0b84911cabe1b57ece33aae3032be47e5f0719ef/sstable/writer.go#L182)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

I found the discussion of ownership a little muddled here. How about:

```
// A dataBlockBuf holds all the state required to compress and write a data block to disk.
// A dataBlockBuf begins its lifecycle owned by the Writer client goroutine. The Writer
// client goroutine adds keys to the sstable, writing directly into a dataBlockBuf's blockWriter
// until the block is full. Once a dataBlockBuf's block is full, the dataBlockBuf may be passed
// to other goroutines for compression and file I/O.
```
</blockquote></details>

Thanks, that's a nicer comment.
___
*[sstable/writer.go, line 185 at r10](https://reviewable.io/reviews/cockroachdb/pebble/1442#-Mtif2yR9bR68fitgjDs:-Mtjh11m3Ww195p_-GCU:b-896fix) ([raw file](https://github.com/cockroachdb/pebble/blob/0b84911cabe1b57ece33aae3032be47e5f0719ef/sstable/writer.go#L185)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

It looks like we can embed the `blockWriter` and avoid the pointer indirection.
</blockquote></details>

Done.

**Review by jbowens — APPROVED** [review]
Nice!

:lgtm:

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1442#-:-Mtn7a2JB564jstEHwdl:b2pss2z)* status: 2 of 3 files reviewed, 2 unresolved discussions (waiting on @itsbilal and @petermattis)

**Comment by jbowens** [comment]
I think a good next step would be to focus on the block-writing/IO side of this change (still without introducing any parallel compression).

**Comment by bananabrick** [comment]
TFTR!
